src/snark/constraints.rs

```diff
@@ -0,0 +1,308 @@
+use ark_ff::{BigInteger, FpParameters, PrimeField};
+use ark_r1cs_std::{
+    alloc::AllocVar, bits::boolean::Boolean, fields::fp::FpVar, ToBitsGadget, ToBytesGadget,
+};
+use ark_relations::{
+    ns,
+    r1cs::{ConstraintSynthesizer, ConstraintSystemRef, Namespace, SynthesisError},
+};
+use core::{borrow::Borrow, marker::PhantomData};
+
+use ark_r1cs_std::alloc::AllocationMode;
+use ark_snark::{CircuitSpecificSetupSNARK, UniversalSetupSNARK, SNARK};
+
+pub trait SNARKGadgets<F: PrimeField, ConstraintF: PrimeField, T: SNARK<F>> {
+    type ProcessedVerifyingKeyVar: AllocVar<T::ProcessedVerifyingKey, ConstraintF> + Clone;
+    type VerifyingKeyVar: AllocVar<T::VerifyingKey, ConstraintF>
+        + ToBytesGadget<ConstraintF>
+        + Clone;
+    type InputVar: AllocVar<Vec<F>, ConstraintF> + FromFieldElementsGadget<F, ConstraintF> + Clone;
+    type ProofVar: AllocVar<T::Proof, ConstraintF> + Clone;
+
+    fn verify_with_processed_vk(
+        cs: ConstraintSystemRef<ConstraintF>,
+        circuit_pvk: &Self::ProcessedVerifyingKeyVar,
+        x: &Self::InputVar,
+        proof: &Self::ProofVar,
+    ) -> Result<Boolean<ConstraintF>, SynthesisError>;
+
+    fn verify(
+        cs: ConstraintSystemRef<ConstraintF>,
+        circuit_vk: &Self::VerifyingKeyVar,
+        x: &Self::InputVar,
+        proof: &Self::ProofVar,
+    ) -> Result<Boolean<ConstraintF>, SynthesisError>;
+}
+
+pub trait CircuitSpecificSetupSNARKGadgets<
+    F: PrimeField,
+    ConstraintF: PrimeField,
+    T: CircuitSpecificSetupSNARK<F>,
+>: SNARKGadgets<F, ConstraintF, T>
+{
+}
+
+pub trait UniversalSetupSNARKGadgets<
+    F: PrimeField,
+    ConstraintF: PrimeField,
+    T: UniversalSetupSNARK<F>,
+>: SNARKGadgets<F, ConstraintF, T>
+{
+    type BoundCircuit: From<T::ComputationBound> + ConstraintSynthesizer<F> + Clone;
+}
+
+pub trait FromFieldElementsGadget<F: PrimeField, ConstraintF: PrimeField>: Sized {
+    fn get_converted_input(src: &Vec<F>) -> Vec<ConstraintF>;
+    fn from_field_elements(
+        cs: ConstraintSystemRef<ConstraintF>,
+        src: &Vec<FpVar<ConstraintF>>,
+    ) -> Result<Self, SynthesisError>;
+}
+
+pub struct BooleanInputVar<F: PrimeField, CF: PrimeField> {
+    pub val: Vec<Vec<Boolean<CF>>>,
+    pub field_phantom: PhantomData<F>,
+    pub constraint_field_phantom: PhantomData<CF>,
```

Comment:
```suggestion
    pub _snark_field_: PhantomData<F>,
    pub _constraint_field: PhantomData<CF>,
```

I would also add a `new` method which constructs this from a `Vec<Vec<Boolean<CF>>>`, so that we can make these fields private.